stand/efi/loader/arch/amd64/Makefile.inc

```diff
@@ -1,15 +1,48 @@
 
+.PATH:	${BOOTSRC}/i386/libi386
+SRCS+=	nullconsole.c \
+	comconsole.c \
+	spinconsole.c
+
+.if ${DO32:U0} == 1 && ${MK_LOADER_IA32} != "no"
+LOADER:=	${LOADER}-ia32
+PROG=	${LOADER}.sym
+
+CFLAGS+=	-I${.CURDIR}/arch/amd64/ia32
+CFLAGS+=	-I${EFISRC}/include/i386
+
+SRCS+= amd64_tramp32.S \
+	start32.S \
+	elf64_freebsd32.c \
+
+NEWVERSWHAT=	"EFI loader" amd64-ia32
+LDSCRIPT=	${.CURDIR}/../loader/arch/amd64/ia32/i386.ldscript
+.if ${LOADER_INTERP} == ${LOADER_DEFAULT_INTERP}
+LINKS=		${BINDIR}/${LOADER}.efi ${BINDIR}/loader-ia32.efi
+.endif
+
+.PATH: ${.CURDIR}/../loader/arch/amd64/ia32
+
+LIBEFI32=		${BOOTOBJ}/efi/libefi32/libefi.a
+
+EFI_TARGET=	efi-app-ia32
+DPADD=		${LDR_INTERP32} ${LIBEFI32} ${LIBSA32} ${LDSCRIPT}
+LDADD=		${LDR_INTERP32} ${LIBEFI32} ${LIBSA32}
+.else
 SRCS+=	amd64_tramp.S \
 	start.S \
 	elf64_freebsd.c \
 	trap.c \
 	multiboot2.c \
 	exc.S
 
-.PATH:	${BOOTSRC}/i386/libi386
-SRCS+=	nullconsole.c \
-	comconsole.c \
-	spinconsole.c
-
 CFLAGS+=	-fPIC
 LDFLAGS+=	-Wl,-znocombreloc
+
+.if ${MK_LOADER_IA32} != "no"
+# Make the 32-bit loader before we start
+.BEGIN: loader32
+loader32: .PHONY
+	${MAKE} -C ${.CURDIR} MAKEOBJDIR=${.OBJDIR}/ia32 DO32=1 ${.TARGETS}
```

Comment:
Why can't we just have a Makefile in the ia32 subdir, use normal SUBDIR+= and have it set DO32 automatically? Explicit recursive make combined with .PHONY is a bit gross.

Reply:
Do you mean a makefile in each `loader_*`?

If so, it would probably look something like this:
`loader_lua/Makefile`
```make
LOADER_INTERP=lua
INSTALL_LOADER_HELP_FILE=no

.if ${MK_LOADER_IA32:Uyes} != "no"
SUBDIR+=        ia32
.endif
  
.include "../loader/Makefile"
```
`loader_lua/ia32/Makefile`
```make
DO32=1
LOADER_INTERP=lua
INSTALL_LOADER_HELP_FILE=no
NEWVERSWHAT="EFI loader" amd64-ia32

.include "../../loader/Makefile"
```
This is an improvement, but the problem here is that loader's makefile expects `${.CURDIR}/../loader` to work and I'm not sure how to get around that properly.

Reply:
I found another way to use `SUBDIR` instead of calling make explicitly.
```
$ cd stand/efi
$ ls
... loader-ia32 loader loader_lua ...
```
```
$ cat loader-ia32/Makefile 

.PATH: ${.CURDIR}/../loader/arch/amd64/ia32

# LOADER_INTERP should be exported by
# the parent.
.if empty(LOADER_INTERP)
.error Expected LOADER_INTERP to be defined.
.endif

DO32=1
NEWVERSWHAT="EFI loader" amd64-ia32
INSTALL_LOADER_HELP_FILE=no

.include "../loader/Makefile"
```
```
$ cat loader/arch/amd64/Makefile.inc 

.PATH:  ${BOOTSRC}/i386/libi386
SRCS+=  nullconsole.c \
        comconsole.c \
        spinconsole.c

CFLAGS+=        -fPIC
LDFLAGS+=       -Wl,-znocombreloc

.if ${DO32:U0} == 0
SRCS+=  amd64_tramp.S \
        start.S \
        elf64_freebsd.c \
        trap.c \
        multiboot2.c \
        exc.S

MAKEOBJDIR:=    ${.OBJDIR}/ia32

.export MAKEOBJDIR
.export LOADER_INTERP

SUBDIR.${MK_LOADER_IA32}+=      ../loader-ia32

.elif ${MK_LOADER_IA32} == "yes"
LOADER:=        ${LOADER}-ia32

CFLAGS+=        -I${EFISRC}/include/i386

LDSCRIPT=       ${.CURDIR}/../loader/arch/amd64/ia32/i386.ldscript
.if ${LOADER_INTERP} == ${LOADER_DEFAULT_INTERP}
LINKS=  ${BINDIR}/${LOADER}.efi ${BINDIR}/loader-ia32.efi
.endif

SRCS+=  amd64_tramp32.S \
        start32.S \
        elf64_freebsd32.c

LIBEFI32=       ${BOOTOBJ}/efi/libefi32/libefi.a
EFI_TARGET=     efi-app-ia32

DPADD=  ${LDR_INTERP32} ${LIBEFI32} ${LIBSA32} ${LDSCRIPT}
LDADD=  ${LDR_INTERP32} ${LIBEFI32} ${LIBSA32}
.endif
```
Unlike the other solution I gave this works, but there's that ugly `MAKEOBJDIR` hack.
Could you elaborate on your initial suggestion? I'm not sure how else to do this.